### PR TITLE
[WIP] feat: support provider aliases

### PR DESCRIPTION
**Pull request by suzuki-shunsuke**
https://github.com/terraform-linters/tflint-ruleset-aws/issues/331

This is still work in progress, and it doesn't work well.

### CODE REVIEW

**Comment by suzuki-shunsuke** [comment]
This seems not to work well.
https://github.com/suzuki-shunsuke/tflint-ruleset-aws/blob/f1695c42740c3d106e9b9b2e37a5ea5f39b3dd32/aws/runner.go#L47-L51

**Comment by suzuki-shunsuke** [comment]
What's the problem?

https://github.com/terraform-linters/tflint-ruleset-aws/runs/5814975813?check_suite_focus=true

```
panic: open aws-sdk-go/models/apis/account/2021-02-01/api-2.json: no such file or directory

goroutine 1 [running]:
main.main()
	/home/runner/work/tflint-ruleset-aws/tflint-ruleset-aws/rules/models/generator/main.go:64 +0xb71
exit status 2
rules/models/models.go:3: running "go": exit status 1
Error: Process completed with exit code 1.
```

File: aws/runner.go [hunk 1]
```diff
@@ -1,41 +1,63 @@
...
 		PluginConfig: config,
-		AwsClient:    client,
+		AwsClients:   clients,
 	}, nil
 }
 
+func (r *Runner) AwsClient(attributes hclext.Attributes) (*Client, error) {
+	provider := "aws"
+	if attr, exists := attributes["provider"]; exists {
+		if err := r.EnsureNoError(r.EvaluateExpr(attr.Expr, &provider, nil), func() error { return nil }); err != nil {
```
Comment (suzuki-shunsuke) [inline aws/runner.go:48]: It doesn't work well.

Comment (suzuki-shunsuke) [reply aws/runner.go:48]: You have to get the alias from the attribute.

e.g.

```tf
resource "aws_instance" "foo" {
  provider = aws.west # You have to get the alias "west".

  # ...
}
```

Comment (suzuki-shunsuke) [reply aws/runner.go:48]: To support Terraform Modules, probably you have to fix something, but unfortunately I don't know well.

```tf
module "aws_vpc" {
  source = "./aws_vpc"
  providers = {
    aws = aws.west
  }
}
```

Comment (wata727) [reply aws/runner.go:49]: > To support Terraform Modules, probably you have to fix something, but unfortunately I don't know well.

Umm, it's a hard problem. To solve this problem, we need to refer to the provider map of the module call.

The `Check()` method of each rule is called for each module (The root module is also one of the modules). So, `GetModuleContent` basically cannot access the parent module containing a module call that calls itself.

The easiest way is to stop support for alias in child modules. If you haven't faced this issue with aliases in modules, you'll be able to accept it.

The best way is to add an API that references the parent module. `GetModuleContent` has options for the module context, imagine something like `ParentModuleCtxType`.
https://github.com/terraform-linters/tflint-plugin-sdk/blob/v0.10.1/tflint/option.go#L18-L19

You should probably be able to achieve the intended behavior by getting the provider map with `Check()` method in the `RuleSet` and then caching it in the Runner.
https://github.com/terraform-linters/tflint-ruleset-aws/blob/7523ed68628e673e2e76572e32634fb5002b5285/aws/ruleset.go#L46-L59

Comment (suzuki-shunsuke) [reply aws/runner.go:49]:
```tf
  provider      = aws.us-east-1
```

```
07:23:43 runner_eval.go:40: [INFO] unevaluable expression found in main.tf:27. TFLint ignores unevaluable expressions.
07:23:43 [ERROR] interceptor/logging.go:18: failed to gRPC request: direction=plugin2host method=/proto.Runner/EvaluateExpr err="rpc error: code = FailedPrecondition desc = unevaluable expression found in main.tf:27"
```

Hmm... It fails to parse the provider attribute.

**Comment by suzuki-shunsuke** [comment]
> > What's the problem?
> > https://github.com/terraform-linters/tflint-ruleset-aws/runs/5814975813?check_suite_focus=true
> 
> It seems to be caused by an update of the aws-sdk-go module. Some rules are generated from the API definition files included in aws-sdk, but the file to be referenced may have changed due to the update.
> 
> If you don't need to update the aws-sdk-go module for this change, I recommend that you revert it or open it as another pull request.

I see. Thank you for your explanation.
`rules/models/aws-sdk-go` was updated automatically, so I committed it.
I reverted. faa5f65 d6e45af

File: aws/runner.go (hunk 1 above)
Comment (wata727) [reply aws/runner.go:49]: `EvaluateExpr` limits the types of expressions that can be evaluated, so cannot evaluate expressions like `aws.*`.
https://github.com/terraform-linters/tflint/blob/v0.35.0/tflint/runner_eval.go#L93-L104

In the Terraform core, it seems that alias doesn't support evaluation, just get it with `gohcl.DecodeExpression`. The same way can be used.
https://github.com/hashicorp/terraform/blob/v1.1.7/internal/configs/provider.go#L58

Comment (suzuki-shunsuke) [reply aws/runner.go:49]: Thank you for your help.

In the Terraform core, resource's "provider" attribute seems to be parsed here.
https://github.com/hashicorp/terraform/blob/3fbedf25430ead97eb42575d344427db3c32d524/internal/configs/resource.go#L136-L140

**Comment by suzuki-shunsuke** [comment]
It is hard to resolve the conflict, so maybe I'll create a new pull request.

**Comment by suzuki-shunsuke** [comment]
https://github.com/terraform-linters/tflint-ruleset-aws/pull/342

**Pull request closed without merging**